Tokenizing the-verdit.txt file

In [6]:
with open("the-verdict.txt", "r", encoding="utf-8") as f:
    raw_text = f.read()

print("Total number of characters: ", len(raw_text))
print(raw_text[:10])

Total number of characters:  20479
I HAD alwa


The Goal is to tokenize 20479-character short story into individual words and special characters, these can later be converted to embeddings for LLM training

Regular expressions are used to split and work on text. To preprocess/tokenize

In [7]:
import re

text = "Hello, world. This is a test."
output = re.split(r'(\s)', text)
print(output)

['Hello,', ' ', 'world.', ' ', 'This', ' ', 'is', ' ', 'a', ' ', 'test.']


In [12]:
output = re.split(r'([,.]|\s)', text)
print(output)

['Hello', ',', '', ' ', 'world', '.', '', ' ', 'This', ' ', 'is', ' ', 'a', ' ', 'test', '.', '']


In [13]:
output = [item for item in output if item.strip()]
print(output)

['Hello', ',', 'world', '.', 'This', 'is', 'a', 'test', '.']


Keeping or removing whitespaces depends on the use case. For codes where spaces matter its not needed to remove spaces.

In [ ]:
# Tokenization scheme as text data has --, ? and other characters that are important

text = "hello, world. Is this-- a test?"
output = re.split(r'([,.-?!_:;()\'"]|--|\s)',text)
output = [ item for item in output if item.strip()]
print(output)

['hello', ',', 'world', '.', 'Is', 'this', '--', 'a', 'test', '?']


In [19]:
# Applying tokenization rules to the raw text data
preprocessed = re.split(r'([,.:;?_!"()\']|--|\s)', raw_text)
preprocessed = [ item.strip() for item in preprocessed if item.strip()]
print(preprocessed[:100])

['I', 'HAD', 'always', 'thought', 'Jack', 'Gisburn', 'rather', 'a', 'cheap', 'genius', '--', 'though', 'a', 'good', 'fellow', 'enough', '--', 'so', 'it', 'was', 'no', 'great', 'surprise', 'to', 'me', 'to', 'hear', 'that', ',', 'in', 'the', 'height', 'of', 'his', 'glory', ',', 'he', 'had', 'dropped', 'his', 'painting', ',', 'married', 'a', 'rich', 'widow', ',', 'and', 'established', 'himself', 'in', 'a', 'villa', 'on', 'the', 'Riviera', '.', '(', 'Though', 'I', 'rather', 'thought', 'it', 'would', 'have', 'been', 'Rome', 'or', 'Florence', '.', ')', '"', 'The', 'height', 'of', 'his', 'glory', '"', '--', 'that', 'was', 'what', 'the', 'women', 'called', 'it', '.', 'I', 'can', 'hear', 'Mrs', '.', 'Gideon', 'Thwing', '--', 'his', 'last', 'Chicago', 'sitter', '--']


In [20]:
print(len(preprocessed))

4690


Converting tokens to token IDs